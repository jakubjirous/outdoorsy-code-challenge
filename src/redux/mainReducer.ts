import { combineReducers } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";

export const mainReducer = combineReducers({
  // Add the generated reducer as a specific top-level slice
  [rentalApi.reducerPath]: rentalApi.reducer,
});
