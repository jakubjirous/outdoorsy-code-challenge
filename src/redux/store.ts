import { configureStore } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";
import { mainReducer } from "./mainReducer";

export const store = configureStore({
  reducer: mainReducer,

  // Adding the api middleware enables caching, invalidation, polling, and other useful features of `rtk-query`.
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(rentalApi.middleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
