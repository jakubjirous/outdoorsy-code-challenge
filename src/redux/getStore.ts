import { configureStore, StoreEnhancer } from "@reduxjs/toolkit";
import { rentalApi } from "./features/rental/rentalApi";
import { mainReducer } from "./mainReducer";
import { RootState, store } from "./store";

/**
 * Gets a store set up with some givens initial state.
 * The store is set up with the RTK Query middleware.
 * Useful mainly for creating redux store provider when writing unit tests.
 *
 * @param initialState
 * @param enhancers
 */
export function getStore(initialState?: RootState, enhancers?: StoreEnhancer[]): typeof store {
  return configureStore({
    reducer: mainReducer,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(rentalApi.middleware),
    enhancers,
  });
}
