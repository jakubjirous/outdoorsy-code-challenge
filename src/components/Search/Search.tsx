import React, { useCallback, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { addKeyword, addKeywords, removeKeyword } from "../../redux/features/search/searchSlice";
import { RootState } from "../../redux/store";
import { KeyboardKey } from "../../utils/keyboardKey";
import { pxToRem } from "../../utils/pxToRem";
import { FaTimes } from "react-icons/fa";

const S = {
  Header: styled.header`
    display: flex;
    flex-flow: row wrap;
    align-items: center;
    gap: var(--spacing-4);
    border: 1px solid var(--color-tertiary);
    border-radius: var(--br);
    padding: var(--spacing-4);
  `,
  Keywords: styled.div`
    display: flex;
    flex-flow: row wrap;
    align-items: center;
    gap: var(--spacing-4);
  `,
  Keyword: styled.div`
    display: flex;
    align-items: center;
    gap: var(--spacing-3);
    border: 1px solid var(--color-primary);
    border-radius: var(--br);
    background-color: var(--color-primary);
    color: var(--white);
    padding: var(--spacing-3);
    padding-left: var(--spacing-4);
  `,
  Input: styled.input`
    min-width: 25%;
    flex: 1;
    border: none;
    border-radius: var(--br);
    outline: none;
    height: ${pxToRem(30)};
  `,
  RemoveButton: styled.button`
    display: flex;
    align-items: center;
    padding: 0;
    border: none;
    background-color: unset;
    cursor: pointer;
    color: white;
  `,
  RemoveIcon: styled(FaTimes)`
    color: var(--white);
    font-size: ${pxToRem(16)};
  `,
};

function Search() {
  const dispatch = useDispatch();

  const keywords = useSelector((state: RootState) => state.search.keywords);

  const [input, setInput] = useState<string>("");

  const [isKeyReleased, setIsKeyReleased] = useState<boolean>(false);

  const onKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLInputElement>) => {
      const { key } = event;
      const trimmedInput = input?.trim();

      // Keywords separation and search submit after hitting the comma on the keyboard
      if (
        (key === KeyboardKey.COMMA || key === KeyboardKey.ENTER) &&
        trimmedInput?.length &&
        !keywords.includes(trimmedInput)
      ) {
        event.preventDefault();
        dispatch(addKeyword(trimmedInput));
        setInput("");
      }

      // Prevent unintentional deletion of all keywords in the input
      if (key === KeyboardKey.BACKSPACE && !input?.length && keywords.length && isKeyReleased) {
        event.preventDefault();
        const keywordsCopy = [...keywords];
        const poppedKeyword = keywordsCopy.pop();
        dispatch(addKeywords(keywordsCopy));
        setInput(poppedKeyword!);
      }

      setIsKeyReleased(false);
    },
    [input, setInput, dispatch, isKeyReleased, setIsKeyReleased]
  );

  const onKeyUp = useCallback(() => {
    setIsKeyReleased(true);
  }, [setIsKeyReleased]);

  const onChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = event.target;
      setInput(value);
    },
    [input, setInput]
  );

  const onRemove = useCallback(
    (index: number) => () => {
      dispatch(removeKeyword(index));
    },
    [dispatch, removeKeyword]
  );

  return (
    <S.Header>
      <S.Keywords>
        {keywords.map((keyword, index) => (
          <S.Keyword key={index} className="keyword">
            {keyword}
            <S.RemoveButton onClick={onRemove(index)}>
              <S.RemoveIcon />
            </S.RemoveButton>
          </S.Keyword>
        ))}
      </S.Keywords>
      <S.Input
        value={input}
        placeholder="Find the perfect RV"
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        onChange={onChange}
        aria-label="Search"
      />
    </S.Header>
  );
}
export default Search;
