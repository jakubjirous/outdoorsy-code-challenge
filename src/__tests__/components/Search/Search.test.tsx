import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import Search from "../../../components/Search/Search";
import { getStore } from "../../../redux/getStore";

const renderContainer = () => {
  return render(
    <Provider store={getStore()}>
      <Search />
    </Provider>
  );
};

describe("<Search/>", () => {
  test("should render correctly", () => {
    const { baseElement } = renderContainer();

    expect(baseElement).toMatchSnapshot();
  });

  test("should add the new keyword to the search after user input and pressing enter key", () => {
    renderContainer();

    const search = screen.getByRole("textbox", {
      name: /search/i,
    });

    userEvent.type(search, "mercedes{enter}");

    expect(screen.getByText(/mercedes/i)).toBeInTheDocument();
  });
});
