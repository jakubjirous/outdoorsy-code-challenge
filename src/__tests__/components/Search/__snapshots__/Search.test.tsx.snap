// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search/> should render correctly 1`] = `
<body>
  <div>
    <header
      class="sc-bczRLJ caFeCX"
    >
      <div
        class="sc-gsnTZi fKrywy"
      />
      <input
        aria-label="Search"
        class="sc-hKMtZM ebRDGl"
        placeholder="Find the perfect RV"
        value=""
      />
    </header>
  </div>
</body>
`;
